vispy/visuals/text/text.py

```diff
@@ -183,27 +166,113 @@ def _text_to_vbo(text, font, anchor_x, anchor_y, lowres_size):
         descender = min(descender, y1 + slop)
         height = max(height, glyph['size'][1] - 2*slop)
 
+    # Get/set the fonts whitespace length and lineheight(size of this ok?)
+    glyph = font[' ']
+    spacewidth = glyph['advance'] * ratio
+    lineheight = height * 1.5
+
+    # Added escape sequences characters: {unicode:offset,...}
+    #       ord('\a') = 7
+    #       ord('\b') = 8
+    #       ord('\f') = 12
+    #       ord('\n') = 10  => linebreak
+    #       ord('\r') = 13
+    #       ord('\t') = 9   => tab, set equal 4 whitespaces?
+    #       ord('\v') = 11  => vertical tab, set equal 4 linebreaks?
+    #   If text coordinate offset > 0 -> it applies to x-direction
+    #   If text coordinate offset < 0 -> it applies to y-direction
+    esc_seq = {7:0, 8:0, 9:-4, 10:1, 11:4, 12:0, 13:0}
+
+    # Keep track of y_offset to set lines at right position
+    y_offset = 0
+
+    # When a linebreaks, record the vertices index value
+    vi_marker = 0
+    ii_offset = 0 # Offset since certain characters won't be drawn
+    """######################################################################"""
+
+    orig_viewport = canvas.context.get_viewport()
+    for ii, char in enumerate(text):
+
+        #NB If the character is an escape sequence, handle it accordingly
+        if ord(char) in esc_seq:
+            if esc_seq[ord(char)] < 0:
+                # Add offset in x-direction
+                x_off += abs(esc_seq[ord(char)]) * spacewidth
+                width += abs(esc_seq[ord(char)]) * spacewidth
+            elif esc_seq[ord(char)] > 0:
+                # Add offset in y-direction, and reset things in x-direction
+                dx = dy = 0
+                # First set things correctly in x-direction
+                #width -= glyph['advance'] * ratio - (glyph['size'][0] - 2*slop)
+
+                if anchor_x == 'right':
+                    dx = -width
+                elif anchor_x == 'center':
+                    dx = -width / 2.
+                vertices['a_position'][vi_marker:vi+4] += (dx, dy)
+                vi_marker = vi+4
```

Comment:
What does this 4 represent?

Reply:
The 4 represent the step size of vertex positions per character of the string. Each character has 4 vertices positions. It needs to be there since the "vi" variable marks the start index of each character I think. Same in the original text.py:
```
186:    vertices['a_position'][vi:vi+4] = position
187:    vertices['a_texcoord'][vi:vi+4] = texcoords
```